doc/sphinx/addendum/generalized-rewriting.rst

```diff
@@ -907,23 +910,25 @@ are applied using the :tacn:`rewrite_strat` tactic.
 :n:`fold @term`
    unify
 
+:n:`fix @ident := @rewstrategy1`
+   fixpoint operator
```

Comment:
Perhaps you could manage at least a sentence here?  These strategies are under-described.  e.g. what the heck is the `ident` for?

Reply:
@jfehrle the construction of fixpoint operators is standard in functional programming.  I don't think it really needs more text, but I'm fine deferring to you on this.  Do you have a wording suggestion that's more clear than
```suggestion
   fixpoint operator locally binding the fixpoint body :n:`@rewstrategy1` to :n:`@ident`, so that :n:`fix @ident := @rewstrategy1` is equivalent to  :n:`@rewstrategy1` with all occurrences of n:`@ident` replaced by :n:`fix @ident := @rewstrategy1`
```

Reply:
Better than nothing, but I've never encountered the concept before, so not quite enough for me to really understand.  Eg I have no idea why/when it is useful or how it would be computed.  Seems to me it would be better to eventually explain the necessary functional programming concepts in the doc or point to a good description.  This is to make Coq  accessible to a broad audience.  I expect there are only a few concepts needed.  WDYT?

Reply:
Here is some existing documentation on the same concept:
- https://en.wikipedia.org/wiki/Fixed-point_combinator (this is probably the clearest / most relevant)
- https://coq.inria.fr/refman/language/core/inductive.html#coq:cmd.Fixpoint
- https://coq.inria.fr/refman/language/core/inductive.html#fixpoint-definitions
- https://en.wikipedia.org/wiki/Fixed_point_(mathematics)

Based on the description in Wikipedia:
> In [combinatory logic](https://en.wikipedia.org/wiki/Combinatory_logic) for [computer science](https://en.wikipedia.org/wiki/Computer_science), a **fixed-point combinator** (or **fixpoint combinator**),[^1] denoted `fix`, is a [higher-order function](https://en.wikipedia.org/wiki/Higher-order_function) (which takes a [function](https://en.wikipedia.org/wiki/Function_(mathematics)) as [argument](https://en.wikipedia.org/wiki/Argument_of_a_function)) that returns some [fixed point](https://en.wikipedia.org/wiki/Fixed_point_(mathematics)) (a value that is mapped to itself) of its argument function, if one exists.
>
> Formally, if the function $f$ has one or more fixed points, then
> $$\textsf{fix}\ f = f\ (\textsf{fix}\ f)\ ,$$
> and hence, by repeated application,
> $$\textsf{fix}\ f = f\ (f\ ( \ldots f\ (\textsf{fix}\ f) \ldots))\ .$$
>
> [^1]: Peyton Jones, Simon L. (1987). *The Implementation of Functional Programming*. Prentice Hall International. [https://www.microsoft.com/en-us/research/uploads/prod/1987/01/slpj-book-1987.pdf](https://www.microsoft.com/en-us/research/uploads/prod/1987/01/slpj-book-1987.pdf), p.26.

Based on this, do you have a suggestion of what to write here?

(Note that the main difference here is that we use substitution instead of application, so that `fix @ident := @rewstrategy1` is equivalent to `@rewstrategy1[@ident ↦ (fix @ident := @rewstrategy1)]` where the syntax `a[b↦c]` means "`a` but with all occurrences of `b` replaced by `c`".  (Substitution tends to be represented either as `a[b/c]` or `a[c/b]` (the choice is inconsistent across papers and books in PL), or more rarely but less ambiguously as `a[b:=c]` or `a[b↦c]`; I'm not sure which the reference manual uses.)  To make the correspondence closer with the wikipedia text, we could imagine writing `fix (@ident := @rewstrategy1)` (not actually valid syntax), where `@ident := @rewstrategy1` is to be read as `fun @ident => @rewstrategy1`.))

> Seems to me it would be better to eventually explain the necessary functional programming concepts in the doc or point to a good description. This is to make Coq accessible to a broad audience. I expect there are only a few concepts needed. WDYT?

Seems reasonable, I'll open an issue.

Reply:
@jfehrle What do you think of the text
```suggestion
   fixpoint operator, where :math:`\texttt{fix }f := v` evaluates to :math:`\subst{v}{f}{(\texttt{fix }f := v)}`
```
(And possibly linking to https://coq.inria.fr/refman/language/cic.html#the-terms for the explanation of `\subst`?)

Or should we just leave it and defer explanation to https://github.com/coq/coq/issues/18416?

@coq/doc-maintainers I don't suppose there's a way to interleave TeX macros like `` :math:`\subst{...}{...}{...}` `` and `` :n:`...` ``, to write ``:math:`\subst{:n:`@rewstrategy1`}{:n:`@ident`}{(:n:`fix @ident := @rewstrategy1`)}` `` or similar?